Userland/Libraries/LibSoftGPU/Device.cpp

```diff
@@ -649,6 +676,66 @@ void Device::draw_primitives(PrimitiveType primitive_type, FloatMatrix4x4 const&
         if (m_clipped_vertices.size() < 3)
             continue;
 
+        if (m_options.lighting_enabled) {
+            auto const& front_material = m_materials.at(0);
+            // Walk through each vertex
+            for (auto& vertex : m_clipped_vertices) {
+                FloatVector4 result_color = { 0.0f, 0.0f, 0.0f, 0.0f };
+
+                // FIXME: We need the ambient scene color available here too!
+                result_color = front_material.emissive + (front_material.ambient * m_lighting_model.scene_ambient_color);
+                for (auto const& light : m_lights) {
+                    if (!light.is_on)
+                        continue;
+
+                    FloatVector4 vertex_to_light;
+
+                    // Light attenuation value.
+                    float light_attenuation_factor = 1.0f;
+                    if (light.position.w() != 0.0f) {
+                        vertex_to_light = light.position - vertex.eye_coordinates;
+                        auto const vertex_to_light_length = vertex_to_light.length();
+                        auto const vertex_to_light_length_squared = vertex_to_light_length * vertex_to_light_length;
+
+                        light_attenuation_factor = 1.0f / (light.constant_attenuation + (light.linear_attenuation * vertex_to_light_length) + (light.quadratic_attenuation * vertex_to_light_length_squared));
+                        vertex_to_light = vertex_to_light / vertex_to_light_length;
+                    } else {
+                        vertex_to_light = light.position.normalized();
+                    }
+
+                    // Spotlight factor
+                    float spotlight_factor = 1.0f;
+                    if (light.spotlight_cutoff_angle != 180.0f) {
+                        const auto spotlight_direction_normalized = light.spotlight_direction.normalized();
+                        const auto light_to_vertex_dot_normalized_spotlight_direction = spotlight_direction_normalized.dot(FloatVector3(vertex_to_light.x(), vertex_to_light.y(), vertex_to_light.z()));
+                        const auto cos_spotlight_cutoff = std::cos(light.spotlight_cutoff_angle * (static_cast<float>(M_PI) / 180.0f));
```

Comment:
Shouldn't we use `AK/Math.h`'s `cos`? Additionally, `glLightf` works with angles in degrees but you might consider converting them to radians when passing the information over to the `Device`. It prevents having to convert those values on the fly and makes most trigonometric calculations easier to read imho.